sw/host/opentitantool/src/command/console.rs

```diff
@@ -0,0 +1,246 @@
+use anyhow::{anyhow, Result};
+use erased_serde::Serialize;
+use nix::unistd::isatty;
+use raw_tty::TtyModeGuard;
+use regex::Regex;
+use std::fs::File;
+use std::io;
+use std::io::{ErrorKind, Read, Write};
+use std::os::unix::io::AsRawFd;
+use std::time::{Duration, Instant};
+use structopt::StructOpt;
+
+use crate::command::CommandDispatch;
+use opentitanlib::io::uart::Uart;
+use opentitanlib::transport::{Capability, Transport};
+use opentitanlib::util::file;
+
+#[derive(Debug, StructOpt)]
+pub struct Console {
+    #[structopt(short, long, help = "Do not print console start end exit messages.")]
+    quiet: bool,
+
+    #[structopt(short, long, help = "Log console output to a file")]
+    logfile: Option<String>,
+
+    #[structopt(short, long, help = "Exit after a timeout in seconds.")]
+    timeout: Option<u64>,
+
+    #[structopt(long, help = "Exit with success if the specified regex is matched.")]
+    exit_success: Option<String>,
+
+    #[structopt(long, help = "Exit with failure if the specified regex is matched.")]
+    exit_failure: Option<String>,
+}
+
+impl CommandDispatch for Console {
+    fn run(&self, transport: &mut dyn Transport) -> Result<Option<Box<dyn Serialize>>> {
+        // We need the UART for the console command to operate.
+        transport.capabilities().request(Capability::UART).ok()?;
+        let mut stdout = std::io::stdout();
+        let mut stdin = std::io::stdin();
+
+        // Set up resources specified by the command line parameters.
+        let logfile = if let Some(filename) = &self.logfile {
+            Some(File::create(filename)?)
+        } else {
+            None
+        };
+        let deadline = if let Some(timeout) = self.timeout {
+            Some(Instant::now() + Duration::new(timeout, 0))
+        } else {
+            None
+        };
+        let exit_success = if let Some(rx) = &self.exit_success {
+            Some(Regex::new(rx)?)
+        } else {
+            None
+        };
+        let exit_failure = if let Some(rx) = &self.exit_failure {
+            Some(Regex::new(rx)?)
+        } else {
+            None
+        };
```

Comment:
I think this section can be simplified to:

```rust
let logfile = self.logfile.as_ref().map(File::create).transpose()?;
let deadline = self.timeout.map(|timeout| Instant::now() + Duration::new(timeout, 0));
let exit_success = self.exit_success.as_ref().map(Regex::new).transpose()?;
let exit_failure = self.exit_failure.as_ref().map(Regex::new).transpose()?;
```

Alternatively, if you don't like using `Option::map`, this can be reduced a bit using `match`:

```rust
let logfile = match &self.logfile {
    Some(filename) => Some(File::create(filename)?),
    None => None,
};
let deadline = match self.timeout {
    Some(timeout) => Some(Instant::now() + Duration::new(timeout, 0),
    None => None,
};
let exit_success = match &self.exit_success {
    Some(rx) => Some(Regex::new(rx)?),
    None => None,
};
let exit_failure = match &self.exit_failure {
    Some(rx) => Some(Regex::new(rx)?),
    None => None,
};
```

Using `match` uses 4 lines per variable rather than 5 lines per variable, and makes the indentation less noisy. Of course this is all down to personal preference.